spec/lib/granite/context_proxy_spec.rb

```diff
@@ -0,0 +1,92 @@
+# spec: unit
+
+require 'granite/context_proxy'
+
+RSpec.describe Granite::ContextProxy do
+  subject { klass.new }
+  let(:klass) do
+    Class.new do
+      include Granite::ContextProxy
+
+      def self.hash
+        'hash'
+      end
+    end
+  end
+  let(:performer) { performers.first }
+  let(:performers) { 2.times.map { instance_double(User) } }
+  let(:context) { {performer: performer} }
+  let(:proxy) { instance_double(Granite::ContextProxy::Proxy) }
+
+  describe '.using' do
+    before do
+      allow(Granite::ContextProxy::Proxy).to receive(:new).with(klass, have_attributes(**context)) { proxy }
+    end
+
+    specify do
+      expect(klass.using(context)).to eq proxy
+    end
+  end
+
+  describe '.as' do
+    before do
+      allow(Granite::ContextProxy::Proxy).to receive(:new).with(klass, have_attributes(**context)) { proxy }
+    end
+
+    specify do
+      expect(klass.as(performer)).to eq proxy
+    end
+  end
+
+  describe '.with_context' do
+    specify do
+      expect { |b| klass.with_context(context, &b) }.to yield_with_no_args
+    end
+
+    it 'sets proxy_content inside block' do
+      klass.with_context(context) do
+        expect(klass.proxy_context).to eq context
+      end
+      expect(klass.proxy_context).to be_nil
+    end
+
+    it 'correctly works with nested contexts' do
+      klass.with_context(context) do
+        expect(klass.proxy_context).to eq context
+        klass.with_context(performer: performers.second) do
+          expect(klass.proxy_context).to eq(performer: performers.second)
+        end
+        expect(klass.proxy_context).to eq context
+      end
+    end
+
+    it 'keeps context in thread safe way' do
+      thread_performers = []
+      threads = 2.times.map do |i|
+        Thread.new do
+          klass.with_context(performer: performers[i]) do
+            Thread.stop
```

Comment:
I am not sure if `Thread.stop` is a barrier. I think Ruby scheduler might invoke this thread back anyway without your control. I used this pattern:

```
    threads = concurrency_level.times.map do |i|
      Thread.new do
        true while wait_for_it
        # logic ...
      end
    end
    wait_for_it = false
    threads.each(&:join)
```

from https://blog.arkency.com/2015/09/testing-race-conditions/

Reply:
https://ruby-doc.org/core-2.7.5/Thread.html#method-c-stop doesn't mention anything about it. So hopefully it doesn't. This pattern won't work for me, since I need to control when thread 1 is resumed and when thread 2 is resumed.

Reply:
@andriusch I see. Check out if you could use an explicit [exchanger](https://ruby-concurrency.github.io/concurrent-ruby/1.1.4/Concurrent/Exchanger.html) for such case. Example [from RES](https://github.com/RailsEventStore/rails_event_store/blob/e7ede7d516355a2b1fa65fe3f4a70393232dd937/ruby_event_store/spec/client_subscriptions_spec.rb#L390-L416)

Reply:
```
irb(main):007:0> Thread.list
=> [#<Thread:0x000056142182aec0 run>, #<Thread:0x0000561421c40500 (irb):2 sleep_forever>]
irb(main):008:0> Thread.new{Thread.stop; puts "FOO" }
=> #<Thread:0x0000561421e38858 (irb):8 run>
irb(main):009:0> Thread.list
=> [#<Thread:0x000056142182aec0 run>, #<Thread:0x0000561421c40500 (irb):2 sleep_forever>, #<Thread:0x0000561421e38858 (irb):8 sleep_forever>]
```

> doesn't mention anything about it. So hopefully it doesn't.

It seems that you are correct.

Reply:
Since we're not really doing any serious thread work, I'd rather not introduce another gem for one spec unless this spec starts causing problems.